clippy_lints/src/single_field_patterns.rs

```diff
@@ -0,0 +1,264 @@
+/// e.g. 1 | (2 | 9) | 3..5
+struct FlatPatterns<'hir, I>
+where
+    I: Iterator<Item = &'hir Pat<'hir>>,
+{
+    patterns: I,
+    stack: Vec<&'hir Pat<'hir>>,
+}
+
+impl<I: Iterator<Item = &'hir Pat<'hir>>> FlatPatterns<'hir, I> {
+    fn new(patterns: I) -> Self {
+        Self {
+            patterns,
+            stack: Vec::new(),
+        }
+    }
+}
+
+impl<I: Iterator<Item = &'hir Pat<'hir>>> Iterator for FlatPatterns<'hir, I> {
+    type Item = &'hir Pat<'hir>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.stack.is_empty() {
+            if let Some(pat) = self.patterns.next() {
+                self.stack.push(pat);
+            } else {
+                return None;
+            }
+        }
+        while let Some(pat) = self.stack.pop() {
+            match pat.kind {
+                PatKind::Or(pats) => self.stack.extend(pats),
+                _ => return Some(pat),
+            }
+        }
+        unreachable!("Or always has 2 patterns, so one of the prior returns must return");
+    }
+}
+
+fn find_sf_lint<'hir>(
+    cx: &LateContext<'_>,
+    patterns: impl Iterator<Item = &'hir Pat<'hir>>,
+    leaf_sf: &impl Fn(&PatKind<'hir>) -> Option<SingleField>,
+) -> Option<(SingleField, Vec<(Span, String)>)> {
+    let fields = FlatPatterns::new(patterns).map(|p| {
+        (
+            p.span,
+            if matches!(p.kind, PatKind::Wild) {
+                Some(SingleField::Unused)
+            } else {
+                leaf_sf(&p.kind)
+            },
+        )
+    });
+    let mut spans = Vec::<(Span, String)>::new();
+    let mut the_one: Option<SingleField> = None;
+    for (target, sf) in fields {
+        if target.from_expansion() {
+            return None;
+        }
+        if let Some(sf) = sf {
+            match sf {
+                SingleField::Unused => {
+                    spans.push((target, String::from("_")));
+                },
+                SingleField::Id { pattern, .. } => {
+                    if let Some(str) = snippet_opt(cx, pattern) {
+                        spans.push((target, str));
+                    } else {
+                        return None;
+                    }
+                    if let Some(one) = the_one {
+                        if sf != one {
+                            return None;
+                        }
+                    } else {
+                        the_one = Some(sf);
+                    }
+                },
+            }
+        } else {
+            return None;
+        }
+    }
+    the_one.map(|one| (one, spans))
+}
+
+fn apply_lint_sf(cx: &LateContext<'_>, span: Span, sugg: impl IntoIterator<Item = (Span, String)>) {
+    span_lint_and_then(
+        cx,
+        SINGLE_FIELD_PATTERNS,
+        span,
+        "this single-variant pattern only matches one field",
+        |diag| {
+            multispan_sugg_with_applicability(diag, "try this", Applicability::MachineApplicable, sugg);
```

Comment:
We/I try to get away from those generic "try this", because they are really unhelpful if the suggestion can't be fully displayed (i.e. multiline span too big).
```suggestion
            multispan_sugg_with_applicability(diag, "try accessing the field directly", Applicability::MachineApplicable, sugg);
```

and

```suggestion
            multispan_sugg_with_applicability(diag, "try this", Applicability::MaybeIncorrect, sugg);
```

because this can lead to rustc warnings (duplicated `_ => {}` match arms).